pkg/meta/model/table.go

```diff
@@ -854,6 +858,96 @@ func (pi *PartitionInfo) GetPartitionIDByName(partitionDefinitionName string) in
 	return -1
 }
 
+// GetDefaultListPartition return the index of Definitions
+// that contains the LIST Default partition otherwise it returns -1
+func (pi *PartitionInfo) GetDefaultListPartition() int {
+	if pi.Type != model.PartitionTypeList {
+		return -1
+	}
+	defs := pi.Definitions
+	for i := range defs {
+		if len(defs[i].InValues) == 0 {
+			return i
+		}
+		for _, vs := range defs[i].InValues {
+			if len(vs) == 1 && vs[0] == "DEFAULT" {
+				return i
+			}
+		}
+	}
+
+	return -1
+}
+
+// CanHaveOverlappingDroppingPartition returns true if special handling is needed
+// during DDL of partitioned tables, where range or list with default partition
+// can have overlapping partitions.
+func (pi *PartitionInfo) CanHaveOverlappingDroppingPartition() bool {
+	if pi.DDLAction == ActionDropTablePartition &&
+		pi.DDLState == StateWriteOnly {
+		return true
+	}
+	return false
+}
+
+// ReplaceWithOverlappingPartitionIdx returns the overlapping partition
+// if there is one and a previous error.
+// Functions based on locatePartitionCommon, like GetPartitionIdxByRow
+// will return the found partition, with an error, since it is being dropped.
+// This function will correct the partition index and error if it can.
+func (pi *PartitionInfo) ReplaceWithOverlappingPartitionIdx(idx int, err error) (int, error) {
+	if err != nil && idx >= 0 {
+		idx = pi.GetOverlappingDroppingPartitionIdx(idx)
+		if idx >= 0 {
+			err = nil
+		}
+	}
+	return idx, err
+}
+
+// GetOverlappingDroppingPartitionIdx takes the index of Definitions
+// and returns possible overlapping partition to use instead.
+// Only used during DROP PARTITION!
+// For RANGE, DROP PARTITION must be a consecutive range of partitions.
+// For LIST, it only takes effect if there is default partition.
+// returns same idx if no overlapping partition
+// return -1 if the partition is being dropped, with no overlapping partition,
+// like for last range partition dropped or no default list partition.
+func (pi *PartitionInfo) GetOverlappingDroppingPartitionIdx(idx int) int {
+	if idx < 0 || idx >= len(pi.Definitions) {
+		return -1
+	}
+	if pi.CanHaveOverlappingDroppingPartition() {
+		switch pi.Type {
+		case model.PartitionTypeRange:
+		DEFINITIONS_LOOP:
+			for i := idx; i < len(pi.Definitions); i++ {
+				for _, def := range pi.DroppingDefinitions {
```

Comment:
Same as the prior comment, better to create a function for this operation like `pi.IsDropping(idx)`